### PR TITLE
Support bendy butt feeds

### PR DESCRIPTION
**Pull request by staltz**
- [x] Supercede #241 
- [x] Code review
- [x] Dependencies that use this branch have stabilised
- [x] Undraft this PR and merge 

@arj03 This branch supercedes https://github.com/ssb-ngi-pointer/ssb-db2/pull/241, for a few reasons: 

- Rebasing was difficult, maybe I did a mistake
- Some dependencies use `publishAs-simple` and it would be dangerous to mutate the branch and make them change
- As soon as this branch is reviewed, we can delete `publishAs-simple`

I also think that this branch could be merged and released as a new version, it still supports existing use cases correctly.

### CODE REVIEW

File: db.js [hunk 1]
```diff
@@ -270,27 +307,33 @@ exports.init = function (sbot, config) {
...
       () => {
         if (content.recps) content = ssbKeys.box(content, content.recps)
-        const latestKVT = state[config.keys.id]
+        const latestKVT = state[keys.id]
         const msgVal = validate.create(
           latestKVT ? { queue: [latestKVT] } : null,
-          config.keys,
-          null,
+          keys,
+          hmacKey,
```
Comment (arj03) [inline db.js:332]: great you fixed this :)

File: db.js [hunk 2]
```diff
@@ -234,10 +274,10 @@ exports.init = function (sbot, config) {
           : null
         validate2.validateSingle(hmacKey, msgVal, latestMsgVal, (err, key) => {
           if (err) return cb(err)
+          updateState({key, value: msgVal})
```
Comment (arj03) [inline db.js:277]: for this to be consistent there needs to be a change in addBatch as well. It only does updateState after log completes. I was thinking about this issue and I think this change is good. Would be nice with a test that shows the problem.

Comment (arj03) [reply db.js:277]: if you want I have give it a go

**Comment by arj03** [comment]
I pushed up a small change to the tests to make them better. I'm keen on having this branch open and changing the repos that use the old branch to this one so we are sure they all work. Especially while bendy butt is still so much in flux. Really appreciate the work you and @mycognosist have been doing on getting that into shape.

File: db.js (hunk 2 above)
Comment (staltz) [reply db.js:277]: Oh, please do, I'm not entirely sure I understood, and a bit tired too

**Comment by staltz** [comment]
> I'm keen on having this branch open and changing the repos that use the old branch to this one so we are sure they all work. Especially while bendy butt is still so much in flux.

Roger that

File: package.json [hunk 3]
```diff
@@ -39,8 +39,10 @@
     "pull-stream": "^3.6.14",
     "push-stream": "^11.0.0",
     "rimraf": "^3.0.2",
+    "ssb-bendy-butt": "~0.12.0",
```
Comment (mycognosist) [inline package.json:42]: Both `ssb-bendy-butt` and `ssb-uri2` have newer versions (`0.12.2` and `1.5.2` respectively). Maybe you'd like to update them here?

**Review by mycognosist — APPROVED** [review]
Everything looks good to me. The only thing I noticed were the outdated versions for a couple of dependencies but other than that I don't notice anything major. Will be awesome to have this merged! :)

**Comment by staltz** [comment]
Thanks for taking a look at the dependencies! :pra

**Comment by staltz** [comment]
Ready for final squirrel wizard @arj03

**Review by arj03 — APPROVED** [review]
Lets do it! I feel really good about this PR now that is has been tested in so many other repos.

**Comment by staltz** [comment]
@arj03 I took a final look at this and found some tiny things to fix still. I'll make a commit

**Comment by staltz** [comment]
@arj03 This last commit does:

- Support bendybutt in addImmediately()
- `add()` just redirects to either a debounced-addBatch() for classic, or addImmediately for bendybutt
- In the case of unknown feed format, we need to do `cb(err)` not `throw err`
- Fixed a typo in one `guardAgainstDuplicateLogs`
- Test the case of unknown feed format

**Comment by arj03** [comment]
Great, less duplicated code. Changes looks good.

**Comment by staltz** [comment]
Wanna press the big green? :)

**Comment by arj03** [comment]
Sure :D

**Pull request merged**